RFC: Establish an RFC process (#2)
lf-lang/rfcs: text/0002-rfc-process.md

active RFCs presents a snapshot of the communities view on the state of the
project and its potential next steps.

## Table of Contents
[Table of Contents]: #table-of-contents

  - [Opening](#lf-rfcs)
  - [Table of Contents]
  - [When you need to follow this process]
  - [Before creating an RFC]
  - [What the process is]
  - [The RFC life-cycle]
  - [Reviewing RFCs]
  - [Implementing an RFC]
  - [RFC Postponement]
  - [License]
  - [Contributions]

## When you need to follow this process
[When you need to follow this process]: #when-you-need-to-follow-this-process

You need to follow this process if you intend to make "substantial" changes to
Lingua Franca, lingo, the reactor runtimes, or the RFC process itself. What
constitutes a "substantial" change is evolving based on community norms and
varies depending on what part of the ecosystem you are proposing to change, but
may include the following.

  - Any semantic or syntactic change to the language that is not a bugfix.
  - Removing language features.
  - Changes to the interface between the compiler and the runtime libraries.
  - Changes to the interfaces between other components of the LF ecosystem, e.g., the language server.

>   - New user facing features in the IDEs and related tools.

Reviewer: ```suggestion
  - New user-facing features in the IDEs and related tools.
```

  - Changes to the graphical syntax of Lingua Franca.

>   - Changes thas require signifcant updates of the handbook.

Reviewer 2: signifcant -> significant

>   - Changes thas require signifcant updates of the handbook.

Reviewer: ```suggestion
  - Changes that require significant updates of the handbook.
```

Some changes do not require an RFC:

>   - Rephrasing, reorganizing, refactoring, or otherwise "changing shape does

Reviewer: ```suggestion
  - Rephrasing, reorganizing, refactoring, or otherwise "changing shape that does
```

    not change meaning".
  - Additions that strictly improve objective, numerical quality criteria
    (warning removal, speedup, better platform coverage, more parallelism, etc.)
  - Additions only likely to be noticed by other developers and invisible to users of Lingua Franca.
  - Trivial changes that can be implemented and discussed in a single PR.

If you submit a pull request to implement a new feature without going through
the RFC process, it may be closed with a polite request to submit an RFC first.

## Before creating an RFC
[Before creating an RFC]: #before-creating-an-rfc

> A hastily-proposed RFC can hurt its chances of acceptance. Low quality

Reviewer: ```suggestion
A hastily-proposed RFC can hurt its chances of acceptance. Low-quality
```

proposals, proposals for previously-rejected features, or those that don't fit

> into the near-term roadmap, may be quickly rejected, which can be demotivating

Reviewer: ```suggestion
proposals, proposals for previously-rejected features, or proposals that don't fit
into the near-term roadmap may be quickly rejected, which can be demotivating
```

for the unprepared contributor. Laying some groundwork ahead of the RFC can
make the process smoother.

Although there is no single way to prepare for submitting an RFC, it is
generally a good idea to pursue feedback from other project developers

> beforehand, to ascertain that the RFC may be desirable; having a consistent

Reviewer: ```suggestion
beforehand to ascertain that the RFC may be desirable. Having a consistent
```

impact on the project requires concerted effort toward consensus-building.
You may talk the idea over on our [official Zulip server](https://lf-lang.zulipchat.com/)
or file issues on this repo for discussion.

As a rule of thumb, receiving encouraging feedback from long-standing project
developers is a good indication that the RFC is worth pursuing.

## What the process is
[What the process is]: #what-the-process-is

In short, to get a major feature added to Lingua Franca, one must first get the

> RFC merged into the RFC repository as a markdown file. At that point the RFC is

Reviewer: ```suggestion
RFC merged into the RFC repository as a Markdown file. At that point the RFC is
```

"active" and may be implemented with the goal of eventual inclusion into Lingua
Franca.

  - Fork the RFC repo [RFC repository]
  - Copy `0000-template.md` to `text/0000-my-feature.md` (where "my-feature" is
    descriptive). Don't assign an RFC number yet; This is going to be the PR

>     number and we'll rename the file accordingly if the RFC is accepted.

Reviewer: ```suggestion
    descriptive). Don't assign an RFC number yet; this is going to be the PR
    number, and we'll rename the file accordingly if the RFC is accepted.
```

  - Fill in the RFC. Put care into the details: RFCs that do not present
    convincing motivation, demonstrate lack of understanding of the design's
    impact, or are disingenuous about the drawbacks or alternatives tend to
    be poorly-received.
  - Submit a pull request (the PR title should start with "RFC:"). As a pull
    request the RFC will receive design feedback from the larger community, and

>     the author should be prepared to revise it in response.

Reviewer 2: I think the important part is that in here is that the PR should have a title beginning with "RFC:". I would not hide this information in brackets (I did not find it for my PR). Maybe:
Submit a pull request titled "RFC: <your feature>

Everything after this is information that might be helpful but since it is part of the same paragraph, it makes the really important information (submit and name the PR) less visible.

  - Now that your RFC has an open pull request, use the issue number of the PR

>     to update your `0000-` prefix to that number.

Reviewer 2: Maybe only:
Update your `0000-` prefix to your PR number.

The rest seems to be a filler.

  - Build consensus and integrate feedback. RFCs that have broad support are
    much more likely to make progress than those that don't receive any
    comments. Feel free to reach out in particular to get help identifying

>     stakeholders and obstacles.

Reviewer 2: It is not clear to me how this will work, and it seems to be overlapping with the next point:
Do we need this as a separate step?

  - The team will discuss the RFC pull request, as much as possible in the
    comment thread of the pull request itself. Offline discussion will be
    summarized on the pull request comment thread.
  - RFCs rarely go through this process unchanged, especially as alternatives
    and drawbacks are shown. You can make edits, big and small, to the RFC to
    clarify or change the design, but make changes as new commits to the pull
    request, and leave a comment on the pull request explaining your changes.
    Specifically, do not squash or rebase commits after they are visible on the

>     pull request.

Reviewer 2: I think this is interesting but it does not fit into the RFC recipe above.

Maybe just "Do not squash or rebase commits in the RFC PR" is enough or move this to the front and add a linebreak after it for details?

  - At some point, a member of the team will propose a "motion for final
    comment period" (FCP), along with a *disposition* for the RFC (merge, close,
    or postpone).
    - This step is taken when enough of the tradeoffs have been discussed and a rough consensus has been reached, so that
      the core team is in a position to make a decision. That does not require
      consensus among all participants in the RFC thread (which is usually
      impossible). However, the argument supporting the disposition on the RFC
      needs to have already been clearly articulated, and there should not be a
      strong consensus *against* that position. Team members use their best
      judgment in taking this step, and the FCP itself ensures there is ample
      time and notification for stakeholders to push back if it is made
      prematurely.
    - For RFCs with lengthy discussion, the motion to FCP is usually preceded by
      a *summary comment* trying to lay out the current state of the discussion
      and major tradeoffs/points of disagreement.
    - The motion for FCP should include a checkbox list that tags major

>       stakeholder and reviewers that need to sign off on the FCP. This is often

Reviewer 2: This sentence is very important for the process and should be more visible. Everything after this seems to be significant but more of an explanation.

      the point at which many team members first review the RFC in full depth.
      Once a majority of reviewers approve (and at most 2 approvals are outstanding),
      the FCP begins.
  - The FCP lasts ten calendar days, so that it is open for at least 5 business
    days. It is also advertised widely,
    e.g. on the LF mailing list. This way all
    stakeholders have a chance to lodge any final objections before a decision

>     is reached.

Reviewer 2: Maybe add this information to the "At some point, a member" FCP part above to have all relevant FCP information in one point.

  - In most cases, the FCP period is quiet, and the RFC is either merged or
    closed. However, sometimes substantial new arguments or ideas are raised,

>     the FCP is canceled, and the RFC goes back into development mode.

Reviewer 2: This seems like something that is nice to know, but is not essential to the process.

## The RFC life-cycle
[The RFC life-cycle]: #the-rfc-life-cycle

Once an RFC becomes "active," anyone may submit an implementation of the
feature as a pull request to the relevant repo. Being "active" is not a rubber
stamp, and in particular still does not mean the feature will ultimately be

> merged; it does mean that in principle all the major stakeholders have agreed

Reviewer: ```suggestion
stamp, and in particular it still does not mean the feature will ultimately be
merged. However, it does mean that in principle, all the major stakeholders have agreed
```

to the feature and are amenable to merging it.

Furthermore, the fact that a given RFC has been accepted and is "active"
implies nothing about what priority is assigned to its implementation, nor does
it imply anything about whether a developer has been assigned the task of
implementing the feature. While it is not *necessary* that the author of the
RFC also write the implementation, it is by far the most effective way to see

> an RFC through to completion: authors should not expect that other project

Reviewer: ```suggestion
an RFC through to completion: Authors should not expect that other project
```

> developers will take on responsibility for implementing their accepted feature.

Reviewer 2: I think
"Once an RFC becomes "active," anyone may submit an implementation of the
feature as a pull request to the relevant repo."
is the important information here. The rest is an explanation of what this means. Do we want to somehow highlight that?

Modifications to "active" RFCs can be done in follow-up pull requests. We
strive to write each RFC in a manner that it will reflect the final design of
the feature; but the nature of the process means that we cannot expect every
merged RFC to actually reflect what the end result will be at the time of the

> next release.

Reviewer 2: Again: "Modifications to "active" RFCs can be done in follow-up pull requests." is the important information that gets mixed with the explanation of what this means.

In general, once accepted, RFCs should not be substantially changed. Only very
minor changes should be submitted as amendments. More substantial changes
should be new RFCs, with a note added to the original RFC. Exactly what counts

> as a "very minor change" is up to the team to decide.

Reviewer 2: This should be part of the previous paragraph. 
Maybe:
"Minor modifications to "active" RFCs can be done in follow-up PRs. Substantial changes require a new RFC."

case a summary from the meeting will be posted back to the RFC pull request.

Final decisions about RFCs are made after the benefits and drawbacks are well
understood. These decisions can be made at any time. When a decision is made,
the RFC pull request will either be merged or closed. In either case, if the
reasoning is not clear from the discussion in the thread, the a comment
describing the rationale for the decision will be added.

## Implementing an RFC
[Implementing an RFC]: #implementing-an-rfc

Some accepted RFCs represent vital features that need to be implemented right
away. Other accepted RFCs can represent features that can wait until some
arbitrary developer feels like doing the work. Every accepted RFC has an
associated issue tracking its implementation in the relevant repository.

The author of an RFC is not obligated to implement it. Of course, the RFC
author (like any other developer) is welcome to post an implementation for
review after the RFC has been accepted.

If you are interested in working on the implementation for an "active" RFC, but
cannot determine if someone else is already working on it, feel free to ask

> (e.g. by leaving a comment on the associated issue).

Reviewer 2: Is this really necessary to document this for our RFC workflow?

I think the interesting part for the implementation would be how issues, PRs, or commits might be associated to an RFC, should the RFC number be used for this? If yes how?

Author: Each active RFC should have a tracking issue (that itself may contain smaller issues/tasks). Commits and PRs that implement the feature completely or provide increments towards a complete implementation should reference the tracking issue. (This isn't any different from how we handle it how, so probably we don't need to mention it explicitly)

# Abstract
[abstract]: #abstract

One paragraph explanation of the feature.

# Motivation
[motivation]: #motivation

Why are we doing this? What use cases does it support? What is the expected outcome?

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Explain the proposal as if it was already included in the language and you were teaching it to another Lingua Franca programmer. That generally means:

- Introducing new named concepts.
- Explaining the feature largely in terms of examples.
- Explaining how programmers should *think* about the feature, and how it should impact the way they use Lingua Franca. It should explain the impact as concretely as possible.
- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
- If applicable, describe the differences between teaching this to existing Lingua Franca programmers and new Lingua Franca programmers.
- Discuss how this impacts the ability to read, understand, and maintain Lingua Franca code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain?

> For implementation-oriented RFCs (e.g. for compiler or runtime internals), this section should focus on how compiler or runtime contributors should think about the change, and give examples of its concrete impact.

Reviewer 2: If I remember the last LF meeting correctly, the main point was that this should be the documentation that is copied in the handbook. Maybe write this instead of:
"Explain the proposal as if it was already included in the language and you were teaching it to another Lingua Franca programmer. That generally means:"

- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?

# Future possibilities
[future-possibilities]: #future-possibilities

Think about what the natural extension and evolution of your proposal would
be and how it would affect the language and project as a whole in a holistic
way. Try to use this section as a tool to more fully consider all possible
interactions with the project and language in your proposal.
Also consider how this all fits into the roadmap for the project
and of the relevant sub-team.

This is also a good place to "dump ideas", if they are out of scope for the
RFC you are writing but otherwise related.

If you have tried and cannot think of any future possibilities,
you may simply state that you cannot think of anything.

Note that having something written down in the future-possibilities section
is not a reason to accept the current or a future RFC; such notes should be
in the section on motivation or rationale in this or subsequent RFCs.

> The section merely provides additional information.

Reviewer 2: I propose to use shorter bullet points:
- How might the feature evolve in the future?

The rest is very wordy and one has to read everything to find things that might be interesting.

---

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The proposed RFC process is heavily inspired by the [Rust
RFCs](https://github.com/rust-lang/rfcs). The description above was adjusted
slightly to better fit our current community structure and scale.

When accepted, the above description will become the content of the main README
file in the repository and define the process we use for proposing, reviewing,
discussing and accepting/rejecting RFCs. We will further add the file called
`0000-template.md` to the root of the repository to act as a template for new
RFCs.

# Drawbacks
[drawbacks]: #drawbacks

- RFCs impose a more formal style of designing and discussion new features,
  which potentially could slow down development.
- RFCs could be interpreted as a strict specification and implemented in a
  water-fall style without considering changing requirements and new insights
  gained from the implementation.
- RFCs focus on the design, not an implementation. Thus, active RFCs will rarely

>   describe the end-result and need to be adjusted when the feature gets

Reviewer: ```suggestion
  describe the end result and may often need to be adjusted when the feature gets
```

  implemented.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

A variation of the proposed RFC process has been used successfully by the Rust
community for the last 10 years. The process is intended to enforce a more
disciplined approach while being as light-weight as possible. The main rationale
for this proposal is to adapt an established and successful RFC process to the
LF community. This proposal is intended as a starting point. Likely, the
proposed process will not fit exactly our needs and can be modified in future
RFCs.

> In order to be light-weight, the proposed process only considers the design of a

Reviewer: ```suggestion
In order to be lightweight, the proposed process only considers the design of a
```

- Python uses [PEP (Python enhancement proposals)](https://peps.python.org/pep-0001/), which is similar to the idea of RFCs but slightly more formal.
- The IETF RFC process is explained [here](https://www.ietf.org/standards/process/informal/).
- Many companies use RFCs internally. For instance Uber has used RFCs sucessfully to scale from a small team to a thousand engineers as described [here](https://blog.pragmaticengineer.com/scaling-engineering-teams-via-writing-things-down-rfcs/).

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Which license should we use for RFCs in this repository?
- How can we send out FCPs to all interested parties? (The mailing list would
  work, but currently there is no way for people to subscribe)

# Future possibilities
[future-possibilities]: #future-possibilities

- The process can be partially automated using a bot. This is done by the Rust
community. However, for now, we should try to establish a process that works for
us (and refine it) before automating the process.

- We can automatically generate a website that lists all active RFCs. The Rust

> community uses mdbook for this.

Reviewer: ```suggestion
community uses mdBook for this.
```

[guide-level-explanation]: #guide-level-explanation

Text to be placed in the README of this repository:

---

[Lingua Franca RFCs]: #lf-rfcs

The RFC (request for comments) process is a framework for managing both
historical and planned changes within the LF community. The record of active
RFCs tells the story of our project, our technology and our thought process. We
believe in an honest and open dialog between all contributors and stakeholders
within the community. The RFC process establishes accountability, helps to
identify and navigate disparate goals, and builds confidence in the direction
the language and the tooling are evolving in.

Many changes, including bug fixes and documentation improvements, can be
implemented and reviewed via the normal GitHub pull request workflow. However,
some changes are more substantial and we ask to put them through a bit of a
design process. In particular, changes that impact other contributors or
stakeholders in the LF community or changes that considerably impact the user
experience. See [When you need to follow this process] for more details.

## TL;DR

Follow these steps to open a new RFC:
1. Fork or clone this repository.

> 2. Copy [`0000-template.md`](0000-template.md) to `text/0000-my-feature.md`,

Reviewer 3: Where does one find `0000-template.md`? This link does not work.

Author: It is not there yet. The nature of this RFC is to propose the content of this file. The template starts [here](https://github.com/lf-lang/rfcs/blob/rfc-process/text/0002-rfc-process.md#abstract-1) in the rendered proposal.

  - New user-facing features in the IDEs and related tools.
  - Changes to the graphical syntax of Lingua Franca.
  - Changes that require significant updates of the handbook.

Some changes do not require an RFC:

  - Rephrasing, reorganizing, refactoring, or otherwise "changing shape that does
    not change meaning".
  - Additions that strictly improve objective, numerical quality criteria
    (warning removal, speedup, better platform coverage, more parallelism, etc.)
  - Additions only likely to be noticed by other developers and invisible to users of Lingua Franca.
  - Trivial changes that can be implemented and discussed in a single PR.

If you submit a pull request to implement a new feature without going through
the RFC process, it may be closed with a polite request to submit an RFC first.

## License
[License]: #license

The contents of this repository are licensed under the 2-Clause BSD License (see [LICENSE](LICENSE)).

## Contributing
[Contributing]: #contributing

Unless explicitly stated otherwise, any contribution intentionally submitted for inclusion in this repository shall be licensed as above, without any additional terms or conditions

---

> Text to be placed in `0000_template.md`:

Reviewer: ```suggestion
Text to be placed in `0000-template.md`:
```

Since we are still a relatively young community a more lightweight process
appears to be appropriate.

Another alternative would be to continue as we did and not establish an RFC
process. This approach, however, will likely not scale to a larger community and
harden the problems we are facing with respect to an overarching design and
architecture.

# Prior art
[prior-art]: #prior-art

- The proposed RFC process is heavily inspired (this is not to say shamelessly copied) from the [Rust community](https://github.com/lf-lang/rfcs)
- Python uses [PEP (Python enhancement proposals)](https://peps.python.org/pep-0001/), which is similar to the idea of RFCs but slightly more formal.
- The IETF RFC process is explained [here](https://www.ietf.org/standards/process/informal/).
- Many companies use RFCs internally. For instance Uber has used RFCs sucessfully to scale from a small team to a thousand engineers as described [here](https://blog.pragmaticengineer.com/scaling-engineering-teams-via-writing-things-down-rfcs/).

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Which license should we use for RFCs in this repository?

> - How can we send out FCPs to all interested parties? (The mailing list would

Reviewer 4: ```suggestion
- How can we send out FCP notifications to all interested parties? (The mailing list would
```
